langchain/prompts/optimized.py

```diff
@@ -0,0 +1,168 @@
+"""Optimized prompt schema definition."""
+import re
+from typing import Any, Callable, Dict, List
+
+from pydantic import BaseModel, Extra, root_validator
+
+from langchain.embeddings.base import Embeddings
+from langchain.prompts.base import DEFAULT_FORMATTER_MAPPING
+from langchain.vectorstores.base import VectorStore
+
+
+class OptimizedPrompt(BaseModel):
+    r"""Schema to represent an optimized prompt for an LLM.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import DynamicPrompt
+            optimized_prompt = OptimizedPrompt(
+                examples=["Say hi. Hi", "Say ho. Ho"],
+                example_separator="\n\n",
+                prefix="",
+                suffix="\n\nSay {foo}"
+                input_variables=["foo"],
+                max_length=200,
+                get_text_length=word_count,
+                vectorstore_client=FAISS.from_texts(examples, OpenAIEmbeddings())
+            )
+    """
+
+    examples: List[str]
+    """A list of the examples that the prompt template expects."""
+
+    example_separator: str = "\n\n"
+    """Example separator, e.g. \n\n, for the dynamic prompt creation."""
+
+    input_variables: List[str] = []
+    """A list of the names of the variables the prompt template expects."""
+
+    prefix: str = ""
+    """Prefix for the prompt."""
+
+    suffix: str = ""
+    """Suffix for the prompt."""
+
+    template_format: str = "f-string"
+    """The format of the prompt template. Options are: 'f-string'."""
+
+    get_text_length: Callable[[str], int] = lambda x: len(re.split("\n| ", x))
+    """Function to measure prompt length. Defaults to word count."""
+
+    max_length: int = 2048
+    """Max length for the prompt, beyond which examples are cut."""
+
+    vectorstore_client: VectorStore
+    """Vectorstore class to use for storing the embeddings."""
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        arbitrary_types_allowed = True
+
+        extra = Extra.forbid
+
+    def template(self, example_list: List[str], **kwargs: Any) -> str:
+        """Return template given full example list."""
+        template = self.example_separator.join(
+            [self.prefix, *example_list, self.suffix]
+        )
+        return DEFAULT_FORMATTER_MAPPING[self.template_format](template, **kwargs)
+
+    def format(self, k: int = 4, **kwargs: Any) -> str:
+        """Optimize the examples in the prompt for the given inputs.
+
+        Args:
+            k: Number of examples to aim for (may be trimmed by optimizer afterwards)
+            kwargs: Any arguments to be passed to the prompt template.
+
+        Returns:
+            A formatted string.
+
+        Example:
+
+        .. code-block:: python
+
+            prompt.format(variable1="foo")
+        """
+        query = " ".join([v for k, v in kwargs.items()])
```

Comment:
hmm this is interesting. im not sure how this will work in practice for prompts with multiple inputs

have you mostly tested this on single input templates? could we try to constrict it to single input templates only?